UID object type not found

```
error[E0022]: UID object type not found: 10
   ┌─ ./sources/factory.move:65:13
   │
65 │     asserts(dynamic_object_field::exists_(pools_id, deny_coin_list_key));
   │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0022]: UID object type not found: 16
   ┌─ ./sources/factory.move:70:26
   │
70 │     let deny_coin_list = dynamic_object_field::borrow<String, factory::DenyCoinList>(pools_id, deny_coin_list_key);
   │                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

exiting with bytecode transformation errors
```

--- a/crates/move-stackless-bytecode/src/dynamic_field_analysis.rs
+++ b/crates/move-stackless-bytecode/src/dynamic_field_analysis.rs
@@ -6,6 +6,7 @@ use crate::{
     function_data_builder::FunctionDataBuilder,
     function_target::{FunctionData, FunctionTarget},
     function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder, FunctionVariant},
+    reaching_def_analysis::{Def, ReachingDefProcessor, ReachingDefState},
     stackless_bytecode::{Bytecode, Operation},
     verification_analysis::get_info,
 };
@@ -178,7 +179,6 @@ pub fn get_function_return_local_pos(local_idx: usize, code: &[Bytecode]) -> Opt
 fn collect_dynamic_field_info(
     targets: &FunctionTargetsHolder,
     builder: &mut FunctionDataBuilder,
-    code: Vec<Bytecode>,
     verified_or_inlined: bool,
 ) -> DynamicFieldInfo {
     let dynamic_field_name_value_fun_qids = vec![
@@ -267,123 +267,135 @@ fn collect_dynamic_field_info(
     .collect_vec();
 
     // compute map of temp index that load object ids to type of object
-    let uid_info = compute_uid_info(&builder.get_target(), targets, &code);
+    let uid_info = compute_uid_info(&builder.get_target(), targets, &builder.data.code);
 
-    let mut info = DynamicFieldInfo::iter_union(code.iter().filter_map(|bc| match bc {
-        Bytecode::Call(_, _, Operation::Function(module_id, fun_id, type_inst), srcs, _) => {
-            let callee_id = module_id.qualified(*fun_id);
+    let alias_info =
+        ReachingDefProcessor::analyze_reaching_definitions(&builder.fun_env, &builder.data);
 
-            let uid_object_type_not_found_error = || {
-                if !verified_or_inlined
-                    && !builder
-                        .fun_env
-                        .get_return_types()
-                        .iter()
-                        .any(|x| x.is_mutable_reference())
-                {
-                    return;
-                }
+    let mut info = DynamicFieldInfo::iter_union(builder.data.code.iter().enumerate().filter_map(
+        |(offset, bc)| match bc {
+            Bytecode::Call(_, _, Operation::Function(module_id, fun_id, type_inst), srcs, _) => {
+                let callee_id = module_id.qualified(*fun_id);
 
-                // TODO: remove this once we don't include modules/functions not included in verification
-                let excluded_modules = vec![
-                    "0x2::dynamic_field",
-                    "0x2::dynamic_object_field",
-                    "0x2::kiosk_extension",
-                ];
-                if excluded_modules
-                    .contains(&builder.fun_env.module_env.get_full_name_str().as_str())
-                {
-                    return;
-                }
+                let uid_object_type_not_found_error = || {
+                    if !verified_or_inlined
+                        && !builder
+                            .fun_env
+                            .get_return_types()
+                            .iter()
+                            .any(|x| x.is_mutable_reference())
+                    {
+                        return;
+                    }
 
-                let loc = builder.get_loc(bc.get_attr_id());
-                builder.fun_env.module_env.env.add_diag(
-                    Diagnostic::new(Severity::Error)
-                        .with_code("E0022")
-                        .with_message(&format!("UID object type not found: {}", srcs[0]))
-                        .with_labels(vec![Label::primary(loc.file_id(), loc.span())]),
-                );
-            };
+                    // TODO: remove this once we don't include modules/functions not included in verification
+                    let excluded_modules = vec![
+                        "0x2::dynamic_field",
+                        "0x2::dynamic_object_field",
+                        "0x2::kiosk_extension",
+                    ];
+                    if excluded_modules
+                        .contains(&builder.fun_env.module_env.get_full_name_str().as_str())
+                    {
+                        return;
+                    }
 
-            if callee_id == builder.fun_env.get_qualified_id() {
-                return None;
-            }
+                    let loc = builder.get_loc(bc.get_attr_id());
+                    builder.fun_env.module_env.env.add_diag(
+                        Diagnostic::new(Severity::Error)
+                            .with_code("E0022")
+                            .with_message(&format!("UID object type not found: {}", srcs[0]))
+                            .with_labels(vec![Label::primary(loc.file_id(), loc.span())]),
+                    );
+                };
 
-            if dynamic_field_name_value_fun_qids.contains(&callee_id) {
-                if let Some((_, obj_type)) = uid_info.get(&srcs[0]) {
-                    return Some(DynamicFieldInfo::singleton(
-                        obj_type.clone(),
-                        NameValueInfo::NameValue {
-                            name: type_inst[0].clone(),
-                            value: type_inst[1].clone(),
-                            is_mut: false,
-                        },
-                    ));
-                } else {
-                    uid_object_type_not_found_error();
-                }
-            }
-
-            if dynamic_field_name_value_fun_mut_qids.contains(&callee_id) {
-                if let Some((_, obj_type)) = uid_info.get(&srcs[0]) {
-                    return Some(DynamicFieldInfo::singleton(
-                        obj_type.clone(),
-                        NameValueInfo::NameValue {
-                            name: type_inst[0].clone(),
-                            value: type_inst[1].clone(),
-                            is_mut: true,
-                        },
-                    ));
-                } else {
-                    uid_object_type_not_found_error();
-                }
-            }
-
-            if dynamic_field_name_only_fun_qids.contains(&callee_id) {
-                if let Some((_, obj_type)) = uid_info.get(&srcs[0]) {
-                    return Some(DynamicFieldInfo::singleton(
-                        obj_type.clone(),
-                        NameValueInfo::NameOnly(type_inst[0].clone()),
-                    ));
-                } else {
-                    uid_object_type_not_found_error();
-                }
-            }
-
-            let fun_id_with_info = targets
-                .get_callee_spec_qid(&builder.fun_env.get_qualified_id(), &callee_id)
-                .unwrap_or(&callee_id);
-
-            // native or intrinsic functions do not access dynamic fields
-            if builder
-                .fun_env
-                .module_env
-                .env
-                .get_function(*fun_id_with_info)
-                .is_native()
-            {
-                return None;
-            }
-
-            let info = match targets.get_data(fun_id_with_info, &FunctionVariant::Baseline) {
-                Some(data) => get_fun_info(data),
-                None => {
-                    // dbg!(&format!(
-                    //     "callee `{}` was filtered out",
-                    //     builder.fun_env.get_full_name_str()
-                    // ));
+                if callee_id == builder.fun_env.get_qualified_id() {
                     return None;
                 }
-            };
 
-            Some(info.instantiate(type_inst))
-        }
-        _ => None,
-    }));
+                if dynamic_field_name_value_fun_qids.contains(&callee_id) {
+                    if let Some((_, obj_type)) =
+                        get_uid_object_type(&uid_info, &alias_info, srcs[0], offset)
+                    {
+                        return Some(DynamicFieldInfo::singleton(
+                            obj_type.clone(),
+                            NameValueInfo::NameValue {
+                                name: type_inst[0].clone(),
+                                value: type_inst[1].clone(),
+                                is_mut: false,
+                            },
+                        ));
+                    } else {
+                        uid_object_type_not_found_error();
+                    }
+                }
+
+                if dynamic_field_name_value_fun_mut_qids.contains(&callee_id) {
+                    if let Some((_, obj_type)) =
+                        get_uid_object_type(&uid_info, &alias_info, srcs[0], offset)
+                    {
+                        return Some(DynamicFieldInfo::singleton(
+                            obj_type.clone(),
+                            NameValueInfo::NameValue {
+                                name: type_inst[0].clone(),
+                                value: type_inst[1].clone(),
+                                is_mut: true,
+                            },
+                        ));
+                    } else {
+                        uid_object_type_not_found_error();
+                    }
+                }
+
+                if dynamic_field_name_only_fun_qids.contains(&callee_id) {
+                    if let Some((_, obj_type)) =
+                        get_uid_object_type(&uid_info, &alias_info, srcs[0], offset)
+                    {
+                        return Some(DynamicFieldInfo::singleton(
+                            obj_type.clone(),
+                            NameValueInfo::NameOnly(type_inst[0].clone()),
+                        ));
+                    } else {
+                        uid_object_type_not_found_error();
+                    }
+                }
+
+                let fun_id_with_info = targets
+                    .get_callee_spec_qid(&builder.fun_env.get_qualified_id(), &callee_id)
+                    .unwrap_or(&callee_id);
+
+                // native or intrinsic functions do not access dynamic fields
+                if builder
+                    .fun_env
+                    .module_env
+                    .env
+                    .get_function(*fun_id_with_info)
+                    .is_native()
+                {
+                    return None;
+                }
+
+                let info = match targets.get_data(fun_id_with_info, &FunctionVariant::Baseline) {
+                    Some(data) => get_fun_info(data),
+                    None => {
+                        // dbg!(&format!(
+                        //     "callee `{}` was filtered out",
+                        //     builder.fun_env.get_full_name_str()
+                        // ));
+                        return None;
+                    }
+                };
+
+                Some(info.instantiate(type_inst))
+            }
+            _ => None,
+        },
+    ));
 
     info.uid_info = uid_info;
 
-    for bc in code {
+    let code = std::mem::take(&mut builder.data.code);
+    for (offset, bc) in code.into_iter().enumerate() {
         match bc {
             Bytecode::Call(
                 attr_id,
@@ -392,7 +404,9 @@ fn collect_dynamic_field_info(
                 mut srcs,
                 aa,
             ) if all_dynamic_field_fun_qids.contains(&module_id.qualified(fun_id)) => {
-                if let Some((obj_local, obj_type)) = info.uid_info.get(&srcs[0]) {
+                if let Some((obj_local, obj_type)) =
+                    get_uid_object_type(&info.uid_info, &alias_info, srcs[0], offset)
+                {
                     srcs[0] = *obj_local;
                     if !dynamic_field_name_value_fun_mut_qids.contains(&module_id.qualified(fun_id))
                         && builder.get_local_type(srcs[0]).is_mutable_reference()
@@ -528,6 +542,25 @@ fn single_uid_field_offset(struct_env: &StructEnv<'_>) -> Option<usize> {
         .ok()
 }
 
+fn get_uid_object_type<'a>(
+    uid_info: &'a BTreeMap<usize, (usize, Type)>,
+    alias_info: &'a BTreeMap<u16, ReachingDefState>,
+    temp_idx: usize,
+    off: usize,
+) -> Option<&'a (usize, Type)> {
+    // First check if temp_idx is directly in uid_info
+    uid_info.get(&temp_idx).or_else(|| {
+        // Otherwise, check if we can find it through alias information
+        alias_info.get(&(off as u16)).and_then(|state| {
+            ReachingDefProcessor::all_aliases(state, &temp_idx)
+                .iter()
+                .filter_map(|alias| uid_info.get(alias))
+                .exactly_one()
+                .ok()
+        })
+    })
+}
+
 pub struct DynamicFieldAnalysisProcessor();
 
 impl DynamicFieldAnalysisProcessor {
@@ -544,7 +577,7 @@ impl FunctionTargetProcessor for DynamicFieldAnalysisProcessor {
         mut data: FunctionData,
         scc_opt: Option<&[FunctionEnv]>,
     ) -> FunctionData {
-        if fun_env.is_intrinsic() {
+        if fun_env.is_native() || fun_env.is_intrinsic() {
             data.annotations.set(DynamicFieldInfo::new(), true);
             return data;
         }
@@ -556,15 +589,9 @@ impl FunctionTargetProcessor for DynamicFieldAnalysisProcessor {
         }
 
         let mut builder = FunctionDataBuilder::new(fun_env, data);
-        let code = std::mem::take(&mut builder.data.code);
 
         // Collect the dynamic field info
-        let info = collect_dynamic_field_info(
-            targets,
-            &mut builder,
-            code,
-            info.verified || info.inlined,
-        );
+        let info = collect_dynamic_field_info(targets, &mut builder, info.verified || info.inlined);
 
         builder
             .data

--- a/crates/move-stackless-bytecode/src/reaching_def_analysis.rs
+++ b/crates/move-stackless-bytecode/src/reaching_def_analysis.rs
@@ -111,7 +111,7 @@ impl ReachingDefProcessor {
 
     /// Compute the set of locals which are borrowed from or which are otherwise used to refer to.
     /// We can't alias such locals to other locals because of reference semantics.
-    fn borrowed_locals(&self, code: &[Bytecode]) -> BTreeSet<TempIndex> {
+    fn borrowed_locals(code: &[Bytecode]) -> BTreeSet<TempIndex> {
         use Bytecode::*;
         code.iter()
             .filter_map(|bc| match bc {
@@ -124,6 +124,55 @@ impl ReachingDefProcessor {
             })
             .collect()
     }
+
+    /// Performs reaching definition analysis and returns the state per instruction.
+    pub fn analyze_reaching_definitions(
+        func_env: &FunctionEnv,
+        data: &FunctionData,
+    ) -> BTreeMap<CodeOffset, ReachingDefState> {
+        let cfg = StacklessControlFlowGraph::new_forward(&data.code);
+        let analyzer = ReachingDefAnalysis {
+            _target: FunctionTarget::new(func_env, &data),
+            borrowed_locals: Self::borrowed_locals(&data.code),
+        };
+        let block_state_map = analyzer.analyze_function(
+            ReachingDefState {
+                map: BTreeMap::new(),
+                havoced: BTreeSet::new(),
+            },
+            &data.code,
+            &cfg,
+        );
+        analyzer.state_per_instruction(block_state_map, &data.code, &cfg, |before, _| {
+            before.clone()
+        })
+    }
+
+    pub fn all_aliases(state: &ReachingDefState, temp_idx: &TempIndex) -> BTreeSet<TempIndex> {
+        let mut visited = BTreeSet::new();
+        let mut to_visit = state
+            .map
+            .get(temp_idx)
+            .unwrap_or(&BTreeSet::new())
+            .iter()
+            .map(|Def::Alias(alias)| *alias)
+            .collect_vec();
+
+        while let Some(current_idx) = to_visit.pop() {
+            if visited.insert(current_idx) {
+                to_visit.extend(
+                    state
+                        .map
+                        .get(&current_idx)
+                        .unwrap_or(&BTreeSet::new())
+                        .iter()
+                        .map(|Def::Alias(alias)| *alias),
+                );
+            }
+        }
+
+        visited
+    }
 }
 
 impl FunctionTargetProcessor for ReachingDefProcessor {
@@ -135,23 +184,7 @@ impl FunctionTargetProcessor for ReachingDefProcessor {
         _scc_opt: Option<&[FunctionEnv]>,
     ) -> FunctionData {
         if !func_env.is_native() {
-            let cfg = StacklessControlFlowGraph::new_forward(&data.code);
-            let analyzer = ReachingDefAnalysis {
-                _target: FunctionTarget::new(func_env, &data),
-                borrowed_locals: self.borrowed_locals(&data.code),
-            };
-            let block_state_map = analyzer.analyze_function(
-                ReachingDefState {
-                    map: BTreeMap::new(),
-                    havoced: BTreeSet::new(),
-                },
-                &data.code,
-                &cfg,
-            );
-            let per_bytecode_state =
-                analyzer.state_per_instruction(block_state_map, &data.code, &cfg, |before, _| {
-                    before.clone()
-                });
+            let per_bytecode_state = Self::analyze_reaching_definitions(func_env, &data);
 
             // Run copy propagation transformation.
             let annotations = ReachingDefAnnotation(per_bytecode_state);

--- a/crates/sui-prover/tests/inputs/dynamic_field/issue-240.move
+++ b/crates/sui-prover/tests/inputs/dynamic_field/issue-240.move
@@ -1,0 +1,27 @@
+module 0x42::foo;
+
+use prover::prover::{requires, asserts};
+
+use std::string::{Self, String};
+use sui::dynamic_field;
+
+public struct Foo has key {
+    id: UID,
+}
+
+public fun borrow_uid(foo: &Foo): &UID {
+    &foo.id
+}
+
+public fun foo(foo: &Foo): bool {
+    *dynamic_field::borrow<String, u64>(&foo.id, string::utf8(b"asdf")) == 10
+}
+
+#[spec(prove)]
+public fun foo_spec(foo: &Foo): bool {
+    let id = borrow_uid(foo);
+    requires(string::try_utf8(b"asdf").is_some());
+    let asdf_key = string::utf8(b"asdf");
+    asserts(dynamic_field::exists_with_type<String, u64>(id, asdf_key));
+    foo(foo)
+}

